### PR TITLE
[WIP] Documentation Refactor 2

### PR DESCRIPTION
**Pull request by fvisin**
This PR resumes and completes #4211.

Thanks @slefrancois for the help rebasing!
@slefrancois @mgermain @nouiz @lamblin @abergeron, I think the current status is decent enough to be merged. Let me know if you have any feedback!

### CODE REVIEW

**Comment by fvisin** [comment]
I pushed some minor changes:
* I added the max numpy/scipy supported versions, `1.11.1` and `0.17.1` respectively.
* I modified the script that generates the docs to prevent it from using GPUs (it would happen when Sphinx loaded theano code). It works locally for me, but double check it if you want (41df69b)

File: doc/install_macos.txt [hunk 1]
```diff
@@ -0,0 +1,155 @@
+.. include:: css.inc
+
+.. _install_macos:
+
+
+Mac OS Installation Instructions
+################################
+
+There are various ways to install Theano dependencies on a Mac. Here
+we describe the process in detail with Canopy, Anaconda, Homebrew or
```
Comment (slefrancois) [inline doc/install_macos.txt]: We no longer support Canopy

File: doc/install_macos.txt [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+add alternative instructions here.
+
+.. _theano-users: http://groups.google.com/group/theano-users?pli=1
+
+.. include:: requirements.txt
+
+.. note::
+
+   If you use version 0.6 or later of Theano, we try to automatically
+   link with the Canopy blas version.  Due to Mac OS peculiarities,
```
Comment (slefrancois) [inline doc/install_macos.txt]: Idem Canopy.

File: doc/scripts/docgen.py [hunk 3]
```diff
@@ -54,6 +54,10 @@ def mkdir(path):
     pythonpath = os.pathsep.join([throot, pythonpath])
     sys.path[0:0] = [throot]  # We must not use os.environ.
 
+    # Make sure we don't use gpu to compile documentation
+    env_th_flags = os.environ['THEANO_FLAGS']
```
Comment (slefrancois) [inline doc/scripts/docgen.py]: This seems to cause an error if the env variable THEANO_FLAGS hasn't been initialized.

**Comment by slefrancois** [comment]
Once the Canopy references are taken out for MacOS, I think it looks good.

Could someone double-check the install instructions before I merge?

**Comment by lamblin** [comment]
I'll have a look.

File: doc/install_macos.txt (hunk 2 above)
Comment (fvisin) [inline doc/install_macos.txt]: RIght, I will remove the note completely.

File: doc/scripts/docgen.py (hunk 3 above)
Comment (fvisin) [inline doc/scripts/docgen.py]: Nice catch, I fixed it.

**Comment by fvisin** [comment]
Thank you @slefrancois for the review. I fixed all the points you raised and pushed the new commits.

I also took some time to update #4141 with a todo list for the installation page. A few points were not addressed, let me know if there are some of them you want me to solve in this PR or if you think we can merge it as is.

File: doc/scripts/docgen.py [hunk 4]
```diff
@@ -99,3 +103,6 @@ def call_sphinx(builder, workdir):
 
     # To go back to the original current directory.
     os.chdir(currentdir)
+
+    # Reset THEANO_FLAGS
+    os.environ['THEANO_FLAGS'] = env_th_flags
```
Comment (lamblin) [inline doc/scripts/docgen.py]: This fail if `THEANO_FLAGS` was not set, in which case `env_th_flags` is `None`.

File: doc/install_generic.inc [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    git checkout origin/v-9998
+    cd libgpuarray
+    </pre></div>
+
+and then follow the `Step-by-step instructions <http://deeplearning.net/software/libgpuarray/installation.html#step-by-step-install>`__.
+
+
+Bleeding-Edge Installation (recommended)
+----------------------------------------
+
```
Comment (lamblin) [inline doc/install_generic.inc]: For the bleeding edge and developer installation, we should have a direct link to the online version at http://deeplearning.net/software/theano_versions/dev/ and a warning to make sure to use the latest instructions.
Otherwise, people may follow earlier, outdated versions of the instructions.

File: doc/install_ubuntu.txt [hunk 6]
```diff
@@ -1,16 +1,23 @@
-:orphan:
```
Comment (lamblin) [inline doc/install_ubuntu.txt]: Why the change of permission (adding executable)?

File: doc/troubleshooting.txt [hunk 7]
```diff
@@ -0,0 +1,368 @@
...
+variable ``http_proxy`` to the proxy address. Using bash this is
+accomplished with the command
+``export http_proxy="http://user:pass@my.site:port/"``
+You can also provide the ``--proxy=[user:pass@]url:port`` parameter
+to pip. The ``[user:pass@]`` portion is optional.
+
+.. _PYTHONPATH:
+
+How to configure the PYTHONPATH
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Comment (lamblin) [inline doc/troubleshooting.txt]: This should not be necessary if Theano was installed through `pip install`, like the updated instructions state.
I would be in favour of removing that section.

File: doc/troubleshooting.txt [hunk 8]
```diff
@@ -0,0 +1,368 @@
...
+    python -c "import numpy; numpy.test()"
+    python -c "import scipy; scipy.test()"
+    THEANO_FLAGS=''; python -c "import theano; theano.test()"
+ 
+All Theano tests should pass (skipped tests and known failures are normal). If
+some test fails on your machine, you are encouraged to tell us what went
+wrong on the ``theano-users@googlegroups.com`` mailing list.
+
+.. warning::
+    Theano's test should **NOT** be run with `device=gpu` or they will fail.
```
Comment (lamblin) [inline doc/troubleshooting.txt]: Mention `device=cuda` first, since the new back-end is supposed to be the default now.

File: doc/troubleshooting.txt [hunk 9]
```diff
@@ -0,0 +1,368 @@
...
+Could speed up and lower memory usage:
+
+- :ref:`cuDNN <libdoc_cuda_dnn>` default cuDNN convolution use less
+   memory then Theano version. But some flags allow it to use more
+   memory. GPU only.
+- Shortly avail, multi-GPU.
+
+Could raise memory usage but speed up computation:
+
+- :attr:`config.lib.cnmem` =1  # Do not raise much memory usage, but if you are at the limit of GPU memory available. GPU only.
```
Comment (lamblin) [inline doc/troubleshooting.txt]: The new back-end has a new way of doing the same thing (`config.gpuarray.preallocate`), we should mention it first.

File: doc/tutorial/using_gpu.txt [hunk 10]
```diff
@@ -1,4 +1,3 @@
-
```
Comment (lamblin) [inline doc/tutorial/using_gpu.txt]: Revert the change of permission here as well.

File: doc/updating.txt [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+    To fix a possible crash, you can clear
+    the Theano cache like this:
+
+    .. code-block:: bash
+
+       theano-cache clear
+
+Bleeding-Edge Installation
+--------------------------
+
```
Comment (lamblin) [inline doc/updating.txt]: Add the warning message about using the latest versions of the instructions and the link to the absolute URL here as well.

File: doc/install_generic.inc (hunk 5 above)
Comment (fvisin) [inline doc/install_generic.inc]: I don't understand what you are referring to: are you talking about the instructions to install libgpuarray or what else? If so, libgpuarray does not have a `dev` installation page (at my best knowledge), hence the detailed git installation instructions to get the right dev commit and the link to the part of the instructions that explains the following steps to compile and install the software.

(Consider these theano instructions are the *dev* ones, as the PR is to the master branch)

File: doc/troubleshooting.txt (hunk 7 above)
Comment (fvisin) [inline doc/troubleshooting.txt]: Agreed!

File: doc/troubleshooting.txt (hunk 8 above)
Comment (fvisin) [inline doc/troubleshooting.txt]: Done. I also specified to use `force_device` to prevent the tests from using the gpu.

Side note: apparently Sphinx does not allow cross-references to `:attr:`, so I cannot link directly to the appropriate section in the configuration page.

Comment (fvisin) [inline doc/troubleshooting.txt]: Edit: nevermind, I found a way! See the latest commit.

File: doc/updating.txt (hunk 11 above)
Comment (fvisin) [inline doc/updating.txt]: Ditto.

File: doc/scripts/docgen.py (hunk 4 above)
Comment (fvisin) [inline doc/scripts/docgen.py]: Fixed.

**Comment by fvisin** [comment]
I incorporated your comments apart from the ones about the link that I didn't understand. Please elaborate on that and check if the new text in troubleshooting is ok (last commit).

File: doc/install_generic.inc (hunk 5 above)
Comment (lamblin) [inline doc/install_generic.inc]: I'm talking about the installation instructions of Theano themselves. I understand that for now they are only going to show up in the dev branch, but when we publish 0.9, then we may have outdated instructions for installing the development version. Or if someone generates the doc locally, and then we change the instructions or requirements for the dev branch.
So I would prefer to include an absolute link to the latest instructions.

**Comment by lamblin** [comment]
Thank for the other fixes.

**Comment by fvisin** [comment]
I added the link in all the installation pages. If that and the troubleshooting section are ok, we should be finally ready to merge!

**Comment by lamblin** [comment]
Damn, that's markdown syntax, not RST...
That should have been the following instead, I think.
```rst
`the latest version of this page <http://deeplearning.net/software/theano_versions/dev/install_windows.html>`_
```

**Comment by lamblin** [comment]
After the link syntax is fixed, I'm OK to merge.

**Comment by fvisin** [comment]
> Damn, that's markdown syntax, not RST...

Oh damn you are right! :) Thanks for the sub-PR.
I merged it and since you were fine with it, I am finally merging this PR to the master!!!

Thanks everybody for this prolonged joint effort! @olimastro @slefrancois @lamblin
